core/src/main/java/org/apache/iceberg/encryption/EnvelopeEncryptionManager.java

```diff
@@ -0,0 +1,152 @@
+   *                               encryption of file streams
+   * @param dataEncryptionConfig configuration for encryption of data files. Future versions will add configuration for
+   *                             encryption of metadata files
+   * @param kmsClient Client of KMS used to wrap/unwrap keys in envelope encryption
+   * @param dataKeyLength If DEKs are not generated by KMS, they will be randomly generated by Iceberg using this length
+   */
+  public EnvelopeEncryptionManager(boolean nativeFormatEncryption,  EnvelopeConfiguration dataEncryptionConfig,
+      KmsClient kmsClient, int dataKeyLength) {
+    if (!nativeFormatEncryption) {
+      throw new UnsupportedOperationException("EnvelopeEncryptionManager currently only supports encryption " +
+          "provided by the underlying file format.");
+    }
+    this.nativeFormatEncryption = nativeFormatEncryption;
+    Preconditions.checkNotNull(dataEncryptionConfig,
+        "Cannot create EnvelopeEncryptionManager because data encryption config is not passed");
+    this.dataEncryptionConfig = dataEncryptionConfig;
+    Preconditions.checkNotNull(dataEncryptionConfig.kekId(),
+        "Cannot create EnvelopeEncryptionManager because table key encryption key ID is not specified");
+    Preconditions.checkNotNull(kmsClient,
+        "Cannot create EnvelopeEncryptionManager because KmsClient is null");
+    this.kmsClient = kmsClient;
+    this.dataKeyLength = dataKeyLength;
+    this.kmsGeneratedKeys = kmsClient.supportsKeyGeneration();
+  }
+
+  @Override
+  public EncryptedOutputFile encrypt(OutputFile rawOutput) {
+    EnvelopeMetadata metadata = generateEnvelopeMetadata(dataEncryptionConfig);
+
+    if (nativeFormatEncryption) {
+      NativeFileCryptoParameters nativeEncryptParams = NativeFileCryptoParameters.create(metadata.dek())
+          .encryptionAlgorithm(metadata.algorithm())
+          .build();
+
+      if (!(rawOutput instanceof NativelyEncryptedFile)) {
+        throw new RuntimeException("Can't natively encrypt " + rawOutput.location() + " because the class " +
+            rawOutput.getClass() + " doesn't implement NativelyEncryptedFile interface");
+      }
+
+      ((NativelyEncryptedFile) rawOutput).setNativeCryptoParameters(nativeEncryptParams);
+
+      return new BaseEncryptedOutputFile(rawOutput, metadata);
+    } else {
+      throw new UnsupportedOperationException("EnvelopeEncryptionManager currently supports only file formats which " +
+          "have native encryption implementations");
+    }
+  }
+
+  @Override
+  public InputFile decrypt(EncryptedInputFile encrypted) {
+    if (encrypted.keyMetadata().buffer() == null) { // unencrypted file
+      return encrypted.encryptedInputFile();
+    }
+    EnvelopeMetadata metadata = EnvelopeMetadataParser.fromJson(encrypted.keyMetadata().buffer());
+    ByteBuffer fileDek = kmsClient.unwrapKey(metadata.wrappedDek(), metadata.kekId());
```

Comment:
I think `unwrap` should only be called when it is sure that decryption is needed to avoid unnecessary RPC calls.